[AIRFLOW-1755] Url prefix support for flower and webservice

gsemet opened the pull request:
Dear Airflow maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-1755


### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:

I would like to submit this change that allows Airflow to support accessing easily the web UI from an endpoint such as `/airflow` and Flower with `/flower`.

For Flower, it is already possible to use the `flowerconfig.py` to convey the `url_prefix` setting, but I prefered aligning it

I propose this change that add 2 new settings on `airflow.cfg` and the CLI arguments:
```
# Root URL to use for the web server
web_server_url_prefix = /airflow
...
# The root URL for Flower
flower_url_prefix = /flower
```

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

I am open to suggestion on how to add unit tests on this PR. Tests will be done on a preproduction cluster running Kubernetes.

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

ashb commented on airflow/contrib/plugins/metastore_browser/templates/metastore_browser/base.html:
```diff
@@ -31,14 +31,14 @@ <h3 style="float: left">
 {{ super() }}
 <link rel="stylesheet" type="text/css" href="{{ url_for("static", filename="dataTables.bootstrap.css") }}">
-<link href="/admin/static/vendor/select2/select2.css" rel="stylesheet">
+<link href="{{ url_for('admin.index')}}static/vendor/select2/select2.css" rel="stylesheet">
```
Should this, (all many similar url_for calls in the templates) not be `{{ url_for("static", filename="vendor/select2/select2.css") }}`?

gsemet replied:
probably, but can I open a ticket on JIRA for someone to take it later? I may not have time to do it on this pull request

msumit replied:
+1 to @ashb.. in the current form the code doesn't look good..

gsemet replied:
It is actually harder than I though, I'll change the ones that are easy but some are tricky (ex: task_instances, what is the right url_for ? ```{{ url_for('admin.task_instances') }}``` does not work)

ashb replied:
When I try `{{ url_for('admin.task_instances') }}` Flask throws a helpful error:

> werkzeug.routing.BuildError: Could not build url for endpoint 'task_instances'. Did you mean 'airflow.task_instances' instead?

And trying that we get the right URL: "/admin/airflow/object/task_instances"

msumit commented on airflow/config_templates/default_airflow.cfg:
```diff
@@ -170,6 +170,9 @@ web_server_host = 0.0.0.0
 web_server_port = 8080
 
+# Root URL to use for the web server
+web_server_url_prefix: /
```
keep the defaults empty.. also it should be `=` insted of `:`?

gsemet replied:
It actually works with both : and =, but I put = to keep coherency

gsemet commented on airflow/www/templates/airflow/dag_details.html:
```diff
@@ -29,7 +29,7 @@
           class="btn"
           style="border: none; background-color:{{ State.color(state)}}; color: {{ State.color_fg(state) }};"
-          href="/admin/taskinstance/?flt0_dag_id_equals={{ dag.dag_id }}&flt2_state_equals={{ state }}">
+          href="{{ url_for('admin.index')}}taskinstance/?flt0_dag_id_equals={{ dag.dag_id }}&flt2_state_equals={{ state }}">
```
This is much difficult than I though:
- ```url_for('admin.taskinstance')}}``` does not work (template render exception)
- ```url_for('airflow.task_instances')}}``` works but the URL pointed is wrong (wants to go to ```/<urlprefix>/admin/airflow/object/taskinstance```) and ends up in a 404

gsemet commented on airflow/www/app.py:
```diff
@@ -30,13 +30,13 @@
-    app = Flask(__name__)
-    app.secret_key = configuration.get('webserver', 'SECRET_KEY')
-    app.config['LOGIN_DISABLED'] = not configuration.getboolean('webserver', 'AUTHENTICATE')
+    app = Flask(__name__, static_url_path=conf.get_url_prefix() + '/static')
```
this fixes static files url

ashb commented on airflow/bin/cli.py:
```diff
@@ -1082,15 +1086,17 @@ def flower(args):
 
-        os.execvp("flower", ['flower', '-b', broka, address, port, api, flower_conf])
+        os.execvp("flower", ['flower', '-b', broka, address, port, api, flower_conf,
+                             url_prefix])
```
I'm not too familiar with this bit of the code. Can you confirm that this url_prefix here maps to FLOWER_URL_PREFIX, and not to WEB_SERVER_URL_PREFIX?

gsemet replied:
Yes, it comes from the line 1611 (`func': flower`). I confirm this works.

I use `/airflow/` in my URL for web service, and `/airflow/flower` for flower UI, so I get:
- `/airflow/admin` for the UI, 
- `/airflow/api` for the experimental REST api, 
- `/airflow/flower` for flower ui)

ashb commented on airflow/bin/cli.py:
```diff
@@ -1066,6 +1066,10 @@ def flower(args):
         api = '--broker_api=' + args.broker_api
 
+    url_prefix = ''
+    if args.url_prefix:
```
I wonder if this should be args.flower_url_prefix?

gsemet replied:
all variables in this methods are related to flower (see some lines above, `api` is not `flower_api`)

ashb commented on airflow/contrib/plugins/metastore_browser/main.py:
```diff
@@ -71,15 +74,17 @@ def table(self):
         return self.render(
-            "metastore_browser/db.html", tables=tables, db=db)
+            "metastore_browser/db.html", tables=tables, db=db,
+            url_prefix=configuration.get_url_prefix())
```
Instead of passing the url_prefix here can we pass the url_prefix to the `app.register_blueprint()` in airflow/www/app.py? If it works that mean that any user plugins will be registered in the correct prefix without having to have to include this in their code too.

gsemet replied:
Yes it is not useful anymore, I removed it

ashb commented on airflow/www/app.py:
```diff
@@ -61,8 +61,10 @@ def create_app(config=None, testing=False):
+            static_url_path=conf.get_url_prefix() + '/admin',
+            index_view=views.HomeView(endpoint='',
+                                      url=conf.get_url_prefix() + '/admin',
+                                      name="DAGs"),
```
I'm not sure what it would break, but I feel that we should pass a `url=conf.get_url_prefix()` call to the Admin() constructor too. This might break everything, but this would map down to `app.register_blueprint(admin, url_prefix=admin.url)` under the hood which feels like the right thing to do.

ashb replied:
Hopefully including this would mean that we don't have to pass url_prefix to all of the render calls though.

gsemet replied:
I just test it it does not change anything :(

ashb commented on docs/integration.rst:
```diff
@@ -1,12 +1,72 @@
+      }
+
+Azure: Microsoft Azure
+----------------------
```
I don' think this header addition should be here

gsemet replied:
indeed

gsemet replied:
I forgot to remove it indeed

ashb commented on docs/integration.rst:
```diff
@@ -1,12 +1,72 @@
+
+    flower_url_prefix = /myorg/flower
+
+Your reverse proxy (ex: nginx) should be configured as follow:
```
Add a note saying that the base_url is  (or whatever the correct setting is named) would be good here too.

gsemet replied:
done

Commit:
[AIRFLOW-1755] URL prefix for flower and webserver

Allows Airflow to support accessing easily the
web UI from an endpoint such as `/airflow`
and Flower with `/flower`.

Adds 2 options in airflow.cfg:
- web_server_url_prefix for the webservice
- flower_url_prefix for Flower

Allow mounting on any kind of HTTP endpoint,
not only FQDN.

Example in`airflow.cfg`:

  # Root URL to use for the web server
  web_server_url_prefix = /airflow
  ...
  # The root URL for Flower
  flower_url_prefix = /flower

Updated documentation as well.

Signed-off-by: Gaetan Semet <gaetan@xeberon.net>

artwr commented on airflow/models.py:
```diff
@@ -1006,6 +1007,7 @@ def mark_success_url(self):
         iso = self.execution_date.isoformat()
         BASE_URL = configuration.get('webserver', 'BASE_URL')
         return BASE_URL + (
+            configuration.get_url_prefix() +
```
Should this be something using `posixpath.join` in case of trailing slashes ?

gsemet replied:
Yes, it requires to aways write a backslash before, so this get url returns either nothing (no prefix) or a path that does not ends with a backslash.
Double slash on a file path is not really a problem, it is more on some web server that might be an issue. But I do not know if I can put a urljoin/ file join everywhere, especially in JavaScript file

artwr commented on airflow/settings.py:
```diff
@@ -84,7 +84,7 @@ def timing(cls, stat, dt):
  """
 
-BASE_LOG_URL = '/admin/airflow/log'
+BASE_LOG_URL = conf.get_url_prefix() + '/admin/airflow/log'
```
Same here. Consider using `posixpath.join`

gsemet replied:
I usually use `os.path.join`, is it better to use `posixpath.join` ?

artwr replied:
While we do not guarantee compatibility with windows per se, os.path.join is os dependent. posixpath.join would be consistent.

artwr reviewed (commented):
You might want to test that the string concatenation does not lead to issues with trailing/leading slashes.
